Commit something (#148)
ghga-de/epic-docs: 76-lynx-boreal/technical_specification.md

specified by the request. If the `UploadContext` is already in the given state, nothing
happens and the UCS returns a successful response.
If user tries to change the status of an `UploadContext` that's already set to `CLOSED`,
they receive an error. Once the update operation is complete, the UCS publishes a Kafka
event reflecting the latest state of the `UploadContext` and returns an HTTP response
indicating the update was successful.

### File Upload Init
The user initiates the upload process for a given single file by making a request to
the `POST /uploads` endpoint. If a valid encrypted access token is supplied with the
request, the UCS ensures it doesn't already have a completed `FileUpload` for the same
file, then adds the `FileUpload` to the associated `UploadContext`. The UCS publishes
upsertion events to Kafka for both the `FileUpload` and `UploadContext`
objects, and finally returns an HTTP response to the user indicating that the file
upload was successfully initiated.

The `ghga-connector` uploads a given file in chunks, and for each chunk it requests
a pre-signed upload URL. If the request includes a valid access token, the UCS

> returns an HTTP response with the pre-signed upload URL if the token is valid.

Author: If caching is done on the connector side, I don't think it needs to be included here. If the UCS makes calls to some service and we implement UCS-side caching for that, then yes.

> returns an HTTP response with the pre-signed upload URL if the token is valid.

Reviewer: Most of it would be on the connector, but we would need to send the correct headers in the UCS response.
But you're right, makes more sense to move that into the epic dealing with the connector.

> ### File Upload Termination

Author: Replaced talk of cancellation with the correct concept.

instructs the UCS to communicate with the S3 instance and terminate (complete) the
multipart upload. The UCS will update the `FileUpload` instance to `COMPLETED` and
publish a Kafka event reflecting the new state. Finally, the UCS will return an HTTP
response indicating the operation was successful and that the file is completely
uploaded.

### File Upload Deletion
The user makes a request to the `DELETE /uploads` endpoint, indicating they wish to
delete a file from the associated Upload Context. If a valid encrypted access token
is supplied with the request, the UCS terminates the ongoing upload if it exists and
deletes the `FileUpload` object from the database. It removes the reference
from the `file_uploads` field in the `UploadContext` and publishes Kafka events
reflecting the deletion of the upload and the new state of the Upload Context.
Finally, the UCS returns an HTTP response to the user indicating the deletion was
successful.

## API Definitions:

### RESTful/Synchronous:

- POST /contexts: Create a new UploadContext
- PATCH /contexts: Update an UploadContext to change the status.
- POST /uploads: Initiate a file upload

> - PATCH /uploads: Terminate an ongoing file upload

Author: Please check out the new wording and let me know if you think it's clear enough or needs more discussion.

> - PATCH /uploads: Terminate an ongoing file upload

Reviewer: That's clear enough

circumstances.

The user then supplies the WPAT to the `ghga-connector` to upload files.
The `ghga-connector` obtains Work Order Tokens (WOTs) automatically by providing the
WPAT to the WPS, and then makes at least two calls for each file:
1. A POST request to the UCS to create the `FileUpload`.
2. A GET request to the UCS to obtain a file part upload URL. This call is repeated for
each file part.

Both the POST and GET requests will require a valid WOT to succeed.

There is an important difference here from the download path's usage of WOTs: during
download, WOTs must carry the file ID being requested for download. For upload, file IDs
are not used, and WOTs merely carry the Upload Context's uuid.
This is because we want the upload process to not be too restrictive nor prescriptive.
This way, the user can upload the files they want without announcing them beforehand.

When the user has completed uploading all files, the user (or a Data Steward) moves the
`UploadContext` to the `LOCKED` state and eventually to the `CLOSED` state by an HTTP
call to the UOS. Once `CLOSED`, the WPS and Claims Repository will process the 'deleted'

> Kafka event and disable/delete any existing claims or tokens.

Author: I think you're right - I've changed it so the claim is kept, but the work package is invalidated on closure. They must create a new work package if the UC is moved from CLOSED to OPEN.

In summary:
- UOS:
  - The Data Steward role required to create `UploadContext`s

>   - A valid WPAT or the Data Steward role is required to change the state of an `UploadContext`.

Author: We don't want to allow Data Stewards to do that too?

>   - A valid WPAT or the Data Steward role is required to change the state of an `UploadContext`.

Reviewer 2: Sorry, let me reorder the sentence 🙈 

For non-data steward users, UOS should only sign a state change request from OPEN to LOCKED.

And yes, for data steward users, all state changes are permitted.

- UCS:
  - Upserting/Deleting `UploadContext` requires an internal service mesh token
  - Getting an upload url requires a valid WOT

For more information on the HTTP API, see the endpoint definitions below.

> ### Work Package Service

Reviewer 3: We can implement it in different ways. If there are operations that should be performed by two different services, we could use a shared key pair and shared WOT schema.

But for cleaner architecture and improved security I think a specific operation should be executed by only one responsible service. And each service should have its own private signing key, and its own WOT schema(s).

The WPS could have the following WOT schemas:

```py
DownloadFileWorkOrder:
   type: "download"
   file_id: str

UploadFileWorkOrder:
   type: "upload" | "delete"
   file_id: str
   context_id: str
   
CreateFileWorkOrder:
  type: "create"
  alias: str
  context_id: str
```  

The DCS would use a DownloadFileWorkOrderContext based on the DownloadFileWorkOrder. If it gets a WOT that has a different type or unexpected properties, it will not validate. Neither will it validate when it gets a WOT with the right schema, but from the UOS instead of the WPS.

Currently, the WOT schema used by the WPS and DCS still contains info on the user, but it is not used and should not be used, because it does not belong in the bounded context of the file services.

Similarly, the UCS would use the UploadFileWorkOrder and CreateFileWorkOrder. I.e. it would use two different auth providers.

The UOS could use the following schemas:

UploadContextWorkOrder:
   type: "lock" | "close"
   context_id: str
   
CreateUploadContextWorkOrder:
  type: "create"
  name: str
```

The key management script will have to create an additional key pair for the UOS.

> ### Work Package Service

Reviewer 3: Each endpoint handling an individual operation would get its individual auth context dependency that would check that the expected schema is passed, signed by the expected service.

This functionality is already included in the service commons library.

> ### Work Package Service

Reviewer 2: I think we agreed that in the context of file uploads we would not issue per-file WOTs but per context. The field `file_id` is also misleading here, because these files are still in a volatile state, have no system-wide id but just a per-context alias.

> ### Work Package Service

Reviewer 3: That's why I separated the CreateFileWorkOrder (where the file id is not known) from the UploadFileWorkOrder (where the file upload has already been initiated).

> ### Work Package Service

Reviewer 3: Yes. It increases security and integrity of the workflow. It does not harm performance since we need these per file (part) WOTs anyway. Cannot harm to do double checks when communicating with external clients.

The check with the claims repository done by the WPS should of course use similar caching mechanisms as we use for the download controller.

4. Add `upload_access_url` to `AccessCheckConfig`
5. Augment the `AccessCheckAdapter` so it can call `/upload-access/users/{user_id}/uploads/{upload_context_id}`
to check if a user has access to a given `UploadContext`
6. Provide a way to distribute WOTs, either by modifying the
`/work-packages/{work_package_id}/files/{file_id}/work-order-tokens` endpoint or
creating a new endpoint

### Claims Repository
Currently, the CRS only manages claims for `datasets`. Claims' visa values for download
access to datasets are a combination of a hardcoded prefix (specific to datasets in
general) and a dataset ID. There are also specific functions to determine if given claim
is for a dataset and if so, which dataset that is - such as `get_dataset_for_value()`.
`create_controlled_access_claim()` should be modified to work for both up- and download
access.

The CRS needs new HTTP endpoints for managing `UploadContext`-specific claims.
Please see the API Definitions for details.

## User Journeys

### UploadContext Creation
A Data Steward makes a call to the UOS to create a new `UploadContext`, specifying the ID(s) of the user(s) who will upload files.

> The UOS calls the UCS's `POST /contexts` endpoint to create the actual `UploadContext` with the state set

Reviewer 2: I would separate the authorization of users from the creation

to `OPEN`. The UCS issues an outbox event, which is consumed by the Work Package
Service. The CRS also gets this outbox event, but ignores it. The CRS only cares if an

> `UploadContext` is *deleted* or *closed*, and in both cases revokes any linked claims.

Reviewer 2: As discussed above, I'm not sure about the linkage between UploadContext state lifecycle and corresponding claims. Re-opening an UploadContext would then require specifying the users with access again?

Author: I'm operating under the notion that a context never moves once it's closed, and that they would open a new context. But I admit that I haven't thought out the ramifications of that

Author: Revised to be able to move from CLOSED to OPEN

Finally, the UOS makes subsequent calls to the CRS to award upload claims to
the user(s) for the given `UploadContext` (specified by ID). Without this claim, the
user cannot create a work package or upload files.

### UploadContext Update
For now, only Data Stewards are allowed to change the state of an `UploadContext`.
Therefore, a user must communicate with a Data Steward when they are ready to LOCK or CLOSE an `UploadContext`.

The Data Steward calls the `PATCH /contexts/{context_id}` endpoint on the UCS API,
authenticating the call with a WOT which also contains the context ID.

> The request body should contain at least the desired context state.

Reviewer 2: Didn't we discuss that these UploadContext modifications would be relayed by UOS rather than requiring a second request from the client? It was my understanding that UOS would sign its own auth token and make a request to UCS directly.

Author: I am not well versed on arbitrary token creation at the moment, but what you say sounds reasonable.

Reviewer 3: See above.

The UCS updates the state of the `UploadContext` to `LOCKED`, `CLOSED`, or `OPEN`, as
specified by the request. If the `UploadContext` is already in the given state, nothing
happens and the UCS returns a successful response.
The initial state of the `UploadContext` is `OPEN`. When the user finishes uploading
files, they can ask a Data Steward to set the Context to a semi-finalized state,
`LOCKED`. It is possible that the user decides they need to make changes, such as

> uploading or removing a file, and in that case a Data Steward can revert the Context to `OPEN`.

Reviewer 2: UOS could allow those state changes for the user directly

the upload status of a single file within an `UploadContext`. Thus, there is a
hierarchical, one-to-many relationship between `UploadContext` and `FileUpload`.

We will define the Pydantic models for these two classes in `ghga-event-schemas`,
along with one stateful config class for each.

#### Inputs
The UCS only receives user input in the form of HTTP requests. It doesn't subscribe to
any Kafka events. However, we will define a slim CLI interface for the service that
exposes commands to `run-rest` and `publish-events`. These commands are commonly seen
across our services at this time.

#### Outputs
There are three categories of output in the UCS: HTTP responses, published events, and
data stored in the database. HTTP responses are described below in the API Definitions
section. The published events and database storage are driven simultaneously by
Hexkit's MongoKafkaDaoPublisher, which the UCS uses to store `UploadContext` and
`FileUpload` instances. Anytime an `UploadContext` or `FileUpload` is created, modified,
or deleted, the UCS publishes a Kafka event containing the latest state. This is done

> according to the Outbox Pattern (not described in further detail here).

Author: Please see [faeabca](https://github.com/ghga-de/epic-docs/pull/148/commits/faeabca30153f50d42aa8e3e4efe8fe6120aee8d)

#### Auth
Before general users (not Data Stewards) can upload files, three things must happen:
1. A Data Steward must create the `UploadContext` via the Upload Orchestration Service.
   - This requires an `CreateUploadContextWorkOrder` token.
   - The Data Steward does not provide a WOT themselves, but the UOS signs and sends
a WOT along when it makes the request to the UCS.
2. A Data Steward must grant the user a claim enabling them to use the `UploadContext`.
3. The user must create a Work Package via the Data Portal to obtain a Work Package
Access Token (WPAT). Only one WPAT is needed for the entire series of files under normal

> circumstances.

Author: The Data Steward accesses the UOS via the Data Portal, so the UOS will be able to inspect the auth context and verify their role.

points to address:
1. Update `WorkPackageRepository` logic to handle CRUD-ing "upload" work packages
2. Revamp Work Order Tokens (see Additional Implementation Details)
3. Listen for outbox events carrying `UploadContext` data, and store/delete the context IDs
   1. For deletions, related work packages should be removed
4. Add `upload_access_url` to `AccessCheckConfig`
5. Augment the `AccessCheckAdapter` so it can call `/upload-access/users/{user_id}/uploads/{context_id}`
to check if a user has access to a given `UploadContext`
6. Provide a way to distribute WOTs, either by modifying the
`/work-packages/{work_package_id}/files/{file_id}/work-order-tokens` endpoint or
creating a new endpoint

### Claims Repository
Currently, the CRS only manages claims for `datasets`. Claims' visa values for download
access to datasets are a combination of a hardcoded prefix (specific to datasets in
general) and a dataset ID. There are also specific functions to determine if given claim
is for a dataset and if so, which dataset that is - such as `get_dataset_for_value()`.
`create_controlled_access_claim()` should be modified to work for both up- and download

> access.

Author: Thanks. I didn't realize the bit about ControlledAccessGrants implying download access (that was not guessable to me anyway). Please see [ddb24b2](https://github.com/ghga-de/epic-docs/pull/148/commits/ddb24b29518e9bc1e40f78906285cfdcfc99a98e) for the new CRS section

## User Journeys

### UploadContext Creation
A Data Steward makes a call to the UOS to create a new `UploadContext`.
The UOS calls the UCS's `POST /contexts` endpoint to create the actual `UploadContext`
with the state set to `OPEN`. The UCS issues an outbox event, which is consumed by the
Work Package Service. The CRS also gets this outbox event, but ignores it. The CRS only
cares if an `UploadContext` is *deleted*, in which case it revokes any linked claims.

> Actually deleting `UploadContext`s is probably a rare occasion, not the norm.

The UOS receives the new `UploadContext` ID from the UCS and returns it to the Data
Steward. The Data Steward then calls the UOS endpoint `POST /access`, supplying the
user ID and context ID, plus any other required data, such as access expiry.

Finally, the UOS makes a call to the CRS to award an upload claim to
the user for the given `UploadContext` (specified by ID). Without this claim, the
user cannot create a work package or upload files.
If the UOS determines that the `UploadContext` does not exist, it returns an error.
It might do this through querying the UCS or by maintaining a record of context IDs in
its own database.

> ### UploadContext Update

Author: The UOS does have such an endpoint. It is defined in the API Definitions section. I will mention that the endpoint can be used to, e.g. change the name/description.

Users are only allowed to make the initial change from `OPEN` to `LOCKED`. Only Data
Stewards may move an `UploadContext` from `LOCKED` to `CLOSED`, `LOCKED` to `OPEN`, or
from `CLOSED` to `OPEN`.

The Data Steward calls the `PATCH /contexts/{context_id}` endpoint on the UOS API,
which verifies their Data Steward role.

> The request body should contain at least the desired context state.

Author: @Cito Storing the date is no problem, but what about the user ID? The current approach uses a WOT that carries no user information, so we could only see whether it was a data steward or not. If we need this, then we need a way to get the user information.

> The request body should contain at least the desired context state.

Reviewer 3: The name and description of the UC, as well as the connection to users, should be only tracked by the UOS and WPS which use normal user authentication. The UCS only needs the ID and state of the UC.

The UCS will update the `FileUpload` instance to `COMPLETED` and publish a Kafka event reflecting the new state.
Finally, the UCS will return an HTTP response indicating the operation was successful.
The Connector displays a message to the user saying the file upload is complete.
If applicable, the Connector proceeds with the next file in the upload batch.

### `FileUpload` Deletion
The user, via the `ghga-connector` and a valid WPAT, makes a request to the
`DELETE /contexts/{context_id}/uploads/{file_id}` endpoint, indicating they wish to
delete a file from the associated Upload Context. If a valid WOT is supplied with the
request, the UCS cancels the ongoing upload if it exists and deletes the `FileUpload`
object from the database. It removes the reference from the `file_uploads` field in the
`UploadContext` and publishes Kafka events reflecting the deletion of the `FileUpload` and the
new state of the Upload Context. Finally, the UCS returns an HTTP response to the user
indicating the deletion was successful.

## API Definitions:

### RESTful/Synchronous:

#### Upload Controller Service:
- `GET /contexts`: Retrieve all `UploadContexts`

>   - Requires WOT

Reviewer 3: Since we agreed to have the UOS track the upload contexts ("boxes"), it should provide an endpoint to GET all  of these boxes, with optional parameters that would allow filtering for user id, names, states and pagination. This endpoint can be called by normal users and data stewards, but when called by normal users will automatically apply the corresponding user id filter.

If the user id filter is provided, the UOS would ask the CRS for all upload grants of that users and return only the boxes with an upload grant for the user.

There should be also a GET endpoint to get an individual box with its aggregated data on both UOS and UCS.

  - Requires WOT
  - Path args and token must agree on context ID and file ID

#### Upload Orchestration Service:
- `POST /contexts`: Create a new `UploadContext` and grant a claim for it for a user
  - Requires Data Steward Role
  - Request body must contain at least one user ID (whether this is one ID or a list of IDs can be decided during implementation)
  - Uses a token to instruct the UCS to create a new `UploadContext` via HTTP call
- `PATCH /contexts/{context_id}`: Update the state of an `UploadContext`
  - Requires Data Steward Role *or* valid claim to the context
    - Only Data Stewards can do `LOCKED` -> `CLOSED` or `LOCKED` -> `OPEN`
    - Users are allowed to do `OPEN` -> `LOCKED`
  - Path arg and token must agree on context ID
  - Request body must contain the new state of the context
  - In turn, calls the matching UCS endpoint
- `POST /access`: Grant user access to an `UploadContext`
  - Requires Data Steward Role
  - Instructs the CRS to create a new claim for each specified user
  - Request body must contain upload context ID and user ID, plus any other req'd info.

>   - Browsing for and revoking claims can be done through the upcoming Claims Browser

Reviewer 3: Offline we decided to let each service own its own version of the upload context tailored to the respective domain (core domain or files). The UOS would have a an object (a "box") with a title, maybe also description and some aggregated info about the uploaded files (number of files, total size) and a state that can be locked. The UCS would have an object ("a bundle") with only an id and an open/closed state. Only the UCS keeps tracks of files. They can be stored in a separate collection linked to the bundle (better not nested, because there can be many files). The bundle object could also keep track of the aggregated data. The UOS subscribes to changes in the UCS bundle in order to update the aggregated data for the box.

#### Work Package Service:
- `GET /users/{user_id}/uploads`: List all `UploadContext` IDs available to the user
- `POST /work-packages/{work_package_id}/uploads/{context_id}/work-order-tokens`: Create a WOT for uploading files

>   - Requires a Work Package Access Token, so the user must have already created a Work Package

Author: That can be in the request body, can't it? The files are not tracked/stored in the WPS for upload like they are with download, so including it in the path doesn't make sense.

#### Claims Repository Service:
- CRS Authentication for upload endpoints should match existing download counterparts
- `GET /upload-access/users/{user_id}/uploads/{context_id}`: check if a user has access to a certain upload context
- `POST /upload-access/users/{user_id}/uploads/{context_id}`: grant upload access
  - This is called by the UOS when the Data Steward grants a user upload access

> - `DELETE /upload-access/users/{user_id}/uploads/{context_id}`: revoke upload access

Author: Done: [3bb03bd](https://github.com/ghga-de/epic-docs/pull/148/commits/3bb03bd59e7d3fbfb85b4f0a0748b5b25c2fca1b)

# Upload Service Intermediary Revision (Lynx Boreal)
**Epic Type:** Implementation Epic

Epic planning and implementation follow the
[Epic Planning and Marathon SOP](https://docs.ghga-dev.de/main/sops/sop001_epic_planning.html).

## Scope
### Outline:
The goal of this epic is to overhaul the Upload Controller Service (UCS) as part of the
new [File Upload concept](https://ghga.pages.hzdr.de/internal.ghga.de/feature_archconcept-file-upload/developer/architecture_concepts/ac007_file_upload/).
While AC 7 includes references to future services like the Study Repository Service,
some of the machinery there is still in the planning phase. This epic will therefore
rely upon some shortcuts, such as extra Data Steward involvement. When the full
architectural concept is realized, the appropriate adaptations will follow in a future
epic.

### Included/Required:

> - Create UOS

Reviewer 3: ```suggestion
- Create new UOS service as described below
```

- Revamp existing UCS logic
- Adapt WPS for "upload" work packages and tokens
- Adapt CRS to also manage claims for upload contexts
- Write Unit and Integration Tests

> - Add `UploadContext` and `FileUpload` schemas to `ghga-event-schemas`

Reviewer 3: ```suggestion
- Add `DataUploadBox`, `FileUploadBundle` and `FileUpload` schemas to `ghga-event-schemas`
```

### Not included:
- Archive test bed integration
- Subsequent FIS or IFRS updates for the upload path
- Any front-end work that may be needed
- GHGA-Connector update

#### New Service: Upload Orchestration Service (UOS)

This epic introduces a new lightweight service to handle authentication concerns while
preserving the boundary between file services and auth concerns like user IDs and roles.

All publicly expose UCS endpoints will require a Work Order Token (WOT), which can only be created from
existing Work Packages in the Work Package Service (WPS) based on claims in the Claims
Repository Service (CRS). We don't want Data Stewards interacting directly with the
CRS to create claims, but the UCS shouldn't be aware of claims or user identities either.
Recognizing that enabling file uploads and creating upload claims are

> bundled actions, we will instead delegate these things to a new service, the UOS.

Reviewer 3: ```suggestion
bundled actions, we will instead delegate these things to a new service, the Upload Orchestration Service (UOS).
```

along with one stateful config class for each.

#### Inputs
The UCS only receives user input in the form of HTTP requests. It doesn't subscribe to
any Kafka events. However, we will define a slim CLI interface for the service that
exposes commands to `run-rest` and `publish-events`. These commands are commonly seen
across our services at this time.

#### Outputs
There are three categories of output in the UCS: HTTP responses, published events, and
data stored in the database. HTTP responses are described below in the API Definitions
section. The published events and database storage are driven simultaneously by
Hexkit's MongoKafkaDaoPublisher, which the UCS uses to store `UploadContext` and
`FileUpload` instances. Anytime an `UploadContext` or `FileUpload` is created, modified,
or deleted, the UCS publishes a Kafka event containing the latest state. This is done
according to the Outbox Pattern (not described in further detail here). 
Other services, like the IFRS and FIS will later consume these events as part of the
official file upload and management strategy. Presently, the WPS will watch
the `UploadContext` events to enable proper work package management.

#### Auth

> **Tokens**  

Reviewer 3: ```suggestion
##### Tokens
```

Service requires just one flavor of WOT to operate, the UCS requires several:
| Token                          | Issuer | Who                  | Action Authorized                                |
|--------------------------------|--------|----------------------|--------------------------------------------------|
| `CreateUploadContextWorkOrder` | UOS    | Data Stewards        | Create a new `UploadContext`                     |
| `ChangeUploadContextWorkOrder` | UOS    | Data Stewards        | Update an existing `UploadContext`               |
| `ViewUploadContextWorkOrder`   | UOS    | Users, Data Stewards | View one or more `UploadContexts`                |
| `CreateFileWorkOrder`          | WPS    | Users                | Initialize a `FileUpload` and obtain new file ID |
| `UploadFileWorkOrder`          | WPS    | Users                | Obtain a part upload URL for a given file ID. Can also be used to delete the file. |

These tokens authorize users to perform the labeled action within the UCS, and they only
carry information necessary for the given action, such as `file_id` or `context_id`.

> **Claims and Work Packages**

Reviewer 3: ```suggestion
##### Claims and Work Packages
```

Before general users (not Data Stewards) can upload files, three things must happen:
1. A Data Steward must create the `UploadContext` via the Upload Orchestration Service.
   - The UOS signs a `CreateUploadContextWorkOrder` token and contacts the UCS.
2. A Data Steward must grant the user a claim enabling them to use the `UploadContext`.

>    - No token is required here, the UOS just verifies the Data Steward role exists.

Reviewer 3: This is a bit misleading. Instead of a WOT token, it uses a "regular" internal auth token.

Author: Are you referring to line 100?

When the user has completed uploading all files, the user contacts the UOS to move the
`UploadContext` to the `LOCKED` state, preventing further changes (Data Stewards can
lock the context as well, if needed). The UOS signs a `ChangeUploadContextWorkOrder`
token and relays the request to the UCS. When the user is
satisfied that all uploads are correct and complete, they ask a Data Steward to move
the `UploadContext` to the `CLOSED` state. Closing the `UploadContext` requires the user
to be a Data Steward. Likewise, only a Data Steward can re-open the context once it
has been `CLOSED`. Claims for closed `UploadContexts` remain valid; the UCS is
responsible for filtering activity based on state.

In summary:
- UOS:
  - Inspects auth context details to discern between Data Stewards and regular users
  - Communicates with the CRS
  - Relays `UploadContext`-related requests to the UCS
  - Signs UOS-owned WOTs for the requests it sends to the UCS.
  - The Data Steward role required to create `UploadContext`s and grant upload claims
  - Data Steward role required for all `UploadContext` changes except `OPEN` -> `LOCKED`
- UCS:

>   - Knows nothing about "users" or "claims"

Reviewer 3: ```suggestion
  - Knows nothing about users , claims, studies, datasets, etc.
```

    INIT = "init"
    COMPLETED = "completed"

class FileUpload(BaseModel):
    """A File Upload"""

    upload_id: UUID4
    state: FileUploadState  # one of INIT, COMPLETED
    alias: str  # the submitted alias from the metadata
    checksum: str
```

## Additional Implementation Details:

### WOT Modifications in WPS
The WPS needs to be able to authorize work in a compartmentalized fashion.
Instead of adapting the existing WOT schema to work for multiple services,
there should be specific schemas dedicated to a given action (work order).
The existing WorkOrderToken model must be modified & renamed.
Accordingly, new WOT schemas should be created in the WPS:

```python
DownloadFileWorkOrder:  # replaces current `WorkOrderToken` model
   type: "download"
   file_id: str  # note: user information is no longer included

CreateUploadContextWorkOrder:
  type: "create"
  name: str

ChangeUploadContextWorkOrder:
   type: "lock" | "close" | "open"
   context_id: str

ViewUploadContextWorkOrder:
   type: "view"
   context_ids: list[str]

CreateFileWorkOrder:
  type: "create"
  alias: str
  context_id: str

UploadFileWorkOrder:
   type: "upload" | "close" | "delete"

>    file_id: str

Reviewer 3: The classes should be ordered by issuer, i.e. first the work orders for files, issued by the WPS, and then the work orders for upload contexts ("boxes") issued by the UOS.

#### Upload Orchestration Service:
- `GET /cases/{case_id}`: Retrieve a `ResearchDataUploadBox` by ID
- `POST /cases`: Create a new `ResearchDataUploadBox`
  - Requires Data Steward Role
  - Signs a `CreateFileUploadBoxWorkOrder` token and sends request to the UCS
  - Returns the ID of the newly created `ResearchDataUploadBox`
- `PATCH /cases/{case_id}`: Update the state of a `ResearchDataUploadBox`
  - Requires Data Steward Role *or* valid claim to the `ResearchDataUploadBox`
    - Only Data Stewards can do `LOCKED` -> `CLOSED` or `LOCKED` -> `OPEN`
    - Users are allowed to do `OPEN` -> `LOCKED`
  - Request body must include the properties to update. Empty body has no effect.
  - Signs an `ChangeFileUploadBoxWorkOrder` token and calls the matching UCS endpoint
    if the request involves toggling the `FileUploadBox`'s mutability
- `POST /access`: Grant user access to a `FileUploadBox`
  - Requires Data Steward Role
  - Instructs the CRS to create a new claim for the specified user
  - Request body must contain:
    - `FileUploadBox` ID
    - User ID
    - IVA ID
    - Any other pertinent information, such as access expiration date.

>   - Browsing for and revoking claims can be done through the upcoming Claims Browser

Author: @Cito When deleting a grant, do we supply the grant ID? If so, does that mean download access grants can also be revoked from the UOS?

>   - Browsing for and revoking claims can be done through the upcoming Claims Browser

Reviewer 3: Yes, we supply the grant ID. However, the CRS has different endpoints for fetching and revoking download and upload grants, so this cannot be mixed up.

>   - Browsing for and revoking claims can be done through the upcoming Claims Browser

Reviewer 3: We could also add restrictions in Istio so that UOS can only talk to the /upload-access path and ARS can only talk to the /download-access path of the CRS if we want to make it even more secure.

#### Work Package Service:
- `GET /users/{user_id}/boxes`: List all `FileUploadBox` IDs available to the user
- `POST /work-packages/{work_package_id}/boxes/{box_id}/work-order-tokens`: Create a WOT for uploading files
  - Requires a Work Package Access Token, so the user must have already created a Work Package
  - The request body must contain the work type and file alias or ID, depending on the
    work type

#### Claims Repository Service:
- CRS Authentication for upload endpoints should match existing download counterparts
- `GET /upload-access/users/{user_id}/boxes`: lists which `FileUploadBoxes` a user can access
- `GET /upload-access/users/{user_id}/boxes/{box_id}`: check if a user has access to a certain upload box
- `POST /upload-access/users/{user_id}/ivas/{iva_id}/boxes/{box_id}`: grant upload access
  - This is called by the UOS when the Data Steward grants a user upload access
- `DELETE /upload-access/users/{user_id}/boxes/{box_id}`: revoke upload access

### Payload Schemas for Events:

```python
class FileUploadBox(BaseModel):
  """A class representing a box that links files belonging to a single dataset"""

  id: UUID4  # unique identifier for the instance

>   mutable: bool  # Whether or not changes to the files in the box are allowed

Author: Opted for `locked`

  size: int  # The total size of all files in the box

class ResearchDataUploadBoxState(StrEnum):
    """The allowed states for an ResearchDataUploadBox instance"""

    OPEN = "open"
    LOCKED = "locked"
    CLOSED = "closed"

class ResearchDataUploadBox(FileUploadBox):
    """A class representing a ResearchDataUploadBox.
    
    Contains all fields from the FileUploadBox and shares IDs.
    """

    state: ResearchDataUploadBoxState  # one of OPEN, LOCKED, CLOSED
    title: str  # short meaningful name for the case
    description: str  # describes the upload case in more detail
    last_changed: UTCDatetime
    changed_by: str  # ID of the user who performed the latest change

class FileUploadState(StrEnum):
    """The allowed states for a FileUpload instance"""

    INIT = "init"
    COMPLETE = "complete"

class FileUpload(BaseModel):
    """A File Upload"""

    upload_id: UUID4
    state: FileUploadState  # one of INIT, COMPLETE
    alias: str  # the submitted alias from the metadata
    checksum: str
    size: int

class AuditRecord(BaseModel):
  """A generic record for audit purposes"""

  created: UTCDatetime
  service: str
  label: str

>   description: str

Author: Added in [6af723c](https://github.com/ghga-de/epic-docs/pull/148/commits/6af723c3ee2b06674df37ccde9518c3f4f292d94)

```

## Additional Implementation Details:

### WOT Modifications in WPS
The WPS needs to be able to authorize work in a compartmentalized fashion.
Instead of adapting the existing WOT schema to work for multiple services,
there should be specific schemas dedicated to a given action (work order).
The existing WorkOrderToken model must be modified & renamed.
The following new WOT schemas should be created in the WPS:

```python
# Authored by the WPS:
DownloadFileWorkOrder:  # replaces current `WorkOrderToken` model
   type: "download"
   file_id: str  # note: user information is no longer included

CreateFileWorkOrder:
  type: "create"
  alias: str
  box_id: str

UploadFileWorkOrder:
   type: "upload" | "close" | "delete"
   file_id: str
   box_id: str

# Authored by the UOS
CreateFileUploadBoxWorkOrder:
  type: "create"
  name: str

ChangeFileUploadBoxWorkOrder:

>    type: "lock" | "close" | "open"

Author: Changed to lock and unlock for a more natural toggle vocabulary

   box_id: str

> ListFilesWorkOrder:

Author: Maybe we should suffix these tokens with WOT instead of WorkOrder; the names are very long.

> ListFilesWorkOrder:

Reviewer 3: Yes, would be shorter, but a bit misleading because the actual WOT as a JWT has a richer schema with issuer, expiration date etc.  Maybe we can cut characters elsewhere? E.g. just UploadBox or FileBox instead of FileUploadBox? Since in this context, it's always a FileUploadBox.

## Scope
### Outline:
The goal of this epic is to overhaul the Upload Controller Service (UCS) as part of the
new [File Upload concept](https://ghga.pages.hzdr.de/internal.ghga.de/feature_archconcept-file-upload/developer/architecture_concepts/ac007_file_upload/).
While this concept includes references to future services like the Study Repository Service,
some of the machinery there is still in the planning phase. This epic will therefore
rely upon some shortcuts, such as extra Data Steward involvement. When the full
architectural concept is realized, the appropriate adaptations will follow in a future
epic.

### Included/Required:
- Implement new Upload Orchestrator Service as described below
- Revamp existing UCS logic
- Adapt WPS for "upload" work packages and tokens
- Adapt CRS to also manage claims for upload boxes
- Add new schemas to `ghga-event-schemas`

### Not included:
- Archive test bed integration
- Subsequent FIS or IFRS updates for the upload path
- Any front-end work that may be needed
- GHGA-Connector update

### Upload Controller Service (UCS)

The UCS is responsible for two things:

> 1. Cataloging individual files for a given study without knowing

Reviewer 2: I would not try to map these concepts to the GHGA domain too much. Especially the assumptions about the relationship to "study" won't hold, so this might be confusing to future readers.

Reviewer 3: I guess the idea was to help understanding by already mentioning the concrete use case and creating a high-level overview that blends the levels, but OTOH you're right, it might be more confusing to readers than helpful. The assumption that the files belong to a study is irrelevant in this context and we might even change that in the future. We should just call it an upload of associated files.

anything about studies or other business concepts.
2. Facilitating the actual upload of said files by:
   - Opening a new multipart S3 upload
   - Dispensing pre-signed part upload URLs
   - Completing/cancelling the S3 upload

> In the UCS, a study is represented by a minimalistic abstraction called a

Reviewer 2: see above

`FileUploadBox`, and a file is represented as a `FileUpload`. The former offers a way to
group file uploads, lock them for changes and check the uniqueness of the included file aliases. The latter offers specific

> file information and a way for other services to be aware of that file's existence.

Reviewer 2: I would not try to map this concepts to the GHGA domain too much. Especially the assumptions about the relationship to "study" won't hold, so this might be confusing to future readers.

Each `FileUploadBox` is labeled by a random UUID and has a boolean that determines
whether any changes can be made to it or a given associated `FileUpload`, as well as
fields for file count and total file size.
Each `FileUpload` has a random UUID, an alias, a file size, a checksum, a bool
indicating if upload is completed, and a field containing its parent `FileUploadBox` ID.
There is a 1:many relationship between `FileUploadBox` and `FileUpload`.

> When there is a new study, a Data Steward creates a new `FileUploadBox`. Then the user can
> create a work package for the new `FileUploadBox` and use the `ghga-connector` to uploa
>  the files for their study. That's the simplified mountaintop view. In reality, users and
> Data Stewards operate the UCS and the `FileUploadBox` indirectly via a service in the core
> domain called UOS, and a corresponding domain object called `ResearchDataUploadBox`.
> See the description of the UOS and the **User Journeys** for more detailed information.

Reviewer 2: ```suggestion
When a user shall be enabled to upload Research Data to GHGA, a Data Steward creates a new `ResearchDataUploadBox` which internally maps to the creation of a `FileUploadBox`. Then the
user can create a work package for the new `FileUploadBox` and use the `ghga-connector` to upload
 Research Data. See the description of the UOS and the **User Journeys** for more detailed information.
```

talks to the Claims Repository Service to control access to the `ResearchDataUploadBox`.
Users/Data Stewards can access the UOS through the Data Portal, and requests are
routed through the auth adapter so the UOS can inspect user IDs and roles.

The other responsibility of the UOS is to relieve the UCS from handling browsing needs.
When anyone wants to view existing `ResearchDataUploadBoxes`, they talk to the UOS, not the UCS.
This is a happy arrangement because viewing this content inherently involves auth:
users should only be able to see a `ResearchDataUploadBox` if they actually have access to it.
Also, the users can be provided with the richer information that is only available in the  `ResearchDataUploadBox`.

How the UOS sits in relation to the UCS:  
When there is a new study, a Data Steward logs on to the Data Portal and creates a new
object, called a `ResearchDataUploadBox`, for uploading the research data. The Data Steward will
enter at least a title and description so they can later identify the box.
The Data Portal will make a call to the UOS with the inputted information.
The UOS will generate a new UUID for the `ResearchDataUploadBox` and then tell the UCS

> to create a corresponding `FileUploadBox` with the same ID. The UOS will return the

Reviewer 2: I fully agree with that, but that does not imply identical IDs imho. And identical IDs are a hidden assumption that is not as explicitly expressed in code as a reference is

> to create a corresponding `FileUploadBox` with the same ID. The UOS will return the

Author: When I first implemented the new UCS I went with identical IDs. However, I will revise that so they are independent.

5. The CRS returns a list of IDs to the UOS.
6. The UOS returns any/all `ResearchDataUploadBoxes`, according to any filtering and pagination applicable.

### `ResearchDataUploadBox` Info Update
> Requires that the user journey "`ResearchDataUploadBox` Creation" has been completed.

![Update Research Data Upload Box](./images/update_rdub.png)

The Data Steward calls the `PATCH /boxes/{box_id}` endpoint on the UOS API
from a page on the Data Portal. The request body should contain the updated 
`ResearchDataUploadBox` or, alternatively, indicate which fields to change and how. This could
be a change to the title or description, for example.
The UOS verifies their Data Steward role is in the auth context and validates the
request parameters. If there are no errors, the UOS updates its copy of the
`ResearchDataUploadBox` and emits an outbox event and returns a successful response.
The UOS will also emit an audit log event with the ID of the user who made the change
as well as the timestamp.

### Work Package Creation
> The user journey "`ResearchDataUploadBox` Creation` must have been completed.

![Create Work Package](./images/create_work_package.png)

> The user uses the Data Portal to see which `ResearchDataUploadBoxes` they have access to, then

Reviewer 3: This is again a bit vague. In principle, both the UOS and the WPS could be used to get the list of boxes of the user. I think in this case, the WPS should be used. We do it similarly for download as well, where the WPS provides the list of datasets. The work package manager in the frontend then has a single service API to talk to. So the user journey should show first the request to the WPS to get the list of boxes to the user, the the request to the WPS to create a work package for the selected box.

Author: I want the diagram to be tight and not redundant. The box retrieval is already diagrammed (better now) and so I have included a short blurb here to direct the reader to that journey. I hope that will suffice.
